mirgecom/artificial_viscosity.py

```diff
@@ -0,0 +1,340 @@
+        dictionary of extra arguments to pass through to the boundary conditions
+
+    Returns
+    -------
+    :class:`mirgecom.fluid.ConservedVars`
+        The artificial viscosity operator applied to *q*.
+    """
+    cv = fluid_state.cv
+    actx = cv.array_context
+    dd_vol = DOFDesc("vol", quadrature_tag)
+    dd_faces = DOFDesc("all_faces", quadrature_tag)
+
+    from warnings import warn
+
+    if boundary_kwargs is not None:
+        warn("The AV boundary_kwargs interface is deprecated, please pass gas_model"
+             " and time directly.")
+        if gas_model is None:
+            gas_model = boundary_kwargs["gas_model"]
+            if "time" in boundary_kwargs:
+                time = boundary_kwargs["time"]
+
+    interp_to_surf_quad = partial(tracepair_with_discr_tag, discr, quadrature_tag)
+
+    def interp_to_vol_quad(u):
+        return op.project(discr, "vol", dd_vol, u)
+
+    if operator_states_quad is None:
+        from mirgecom.gas_model import make_operator_fluid_states
+        operator_states_quad = make_operator_fluid_states(
+            discr, fluid_state, gas_model, boundaries, quadrature_tag)
+
+    vol_state_quad, inter_elem_bnd_states_quad, domain_bnd_states_quad = \
+        operator_states_quad
+
+    # Get smoothness indicator based on mass component
+    if indicator is None:
+        indicator = smoothness_indicator(discr, fluid_state.mass_density,
+                                         kappa=kappa, s0=s0)
+
+    if grad_cv is None:
+        from mirgecom.navierstokes import grad_cv_operator
+        grad_cv = grad_cv_operator(discr, gas_model, boundaries, fluid_state,
+                                   time=time, quadrature_tag=quadrature_tag,
+                                   operator_states_quad=operator_states_quad)
+
+    # Compute R = alpha*grad(Q)
+    r = -alpha * indicator * grad_cv
+
+    def central_flux_div(utpair):
+        dd = utpair.dd
+        normal = thaw(actx, discr.normal(dd))
+        return op.project(discr, dd, dd.with_dtag("all_faces"),
+                          # This uses a central vector flux along nhat:
+                          # flux = 1/2 * (grad(Q)- + grad(Q)+) .dot. nhat
+                          divergence_numerical_flux(utpair.int, utpair.ext)@normal)
+
+    # Total flux of grad(Q) across element boundaries
+    r_bnd = (
+        # Rank-local and cross-rank (across parallel partitions) contributions
+        + sum(central_flux_div(interp_to_surf_quad(tpair=tpair))
+              for tpair in interior_trace_pairs(discr, r, tag=_AVRTag))
+
+        # Contributions from boundary fluxes
+        + sum(boundaries[btag].av_flux(
+            discr,
+            btag=as_dofdesc(btag).with_discr_tag(quadrature_tag),
+            diffusion=r) for btag in boundaries)
+    )
+
+    # Return the AV RHS term
+    return -div_operator(discr, dd_vol, dd_faces, interp_to_vol_quad(r), r_bnd)
+
+
+def smoothness_indicator(discr, u, kappa=1.0, s0=-6.0):
```

Comment:
I think this can be built simply out of the modal discretization facility in meshmode. There's definitely no need for this to make use of a loopy kernel. @kaushikcfd had *also* worked on a more lazy-compatible version of this. I am surprised to see that that is not part of this.

Reply:
The tagging bit won't work, but the einsum bit should be fine. I don't have a preference for which PR the einsum is introduced in.

Reply:
Do we have a path forward for this item?

Reply:
fwiw, here's the error i get when trying to run with @kaushikcfd indicator:

```
    indicator = smoothness_indicator(discr, fluid_state.mass_density,
  File "/Users/mtcampbe/CEESD/devel/tau/main/emirge/mirgecom/mirgecom/artificial_viscosity.py", line 318, in smoothness_indicator
    from meshmode.transform_metadata import DiscretizationDOFAxisTag
ImportError: cannot import name 'DiscretizationDOFAxisTag' from 'meshmode.transform_metadata' (/Users/mtcampbe/CEESD/devel/tau/
```

Reply:
The `actx.tag_axis` part would be illegal, but without that the call to `actx.einsum` should be legal on `meshmode` main. I'm ok with the rewrite to `einsum` either here or in #666.

Reply:
I just added @kaushikcfd version of this in ([732f005](https://github.com/illinois-ceesd/mirgecom/pull/582/commits/732f005b632bd303fb8bbc61669a95193ce1379e)).   The tests passed locally with main (:tada:).